src/main/java/pl/stormit/tinyurl/controller/UrlController.java

```diff
@@ -38,9 +39,13 @@ public List<Url> getUrls() {
 
     @GetMapping("/{shortUrl}")
     public ResponseEntity<UrlDto> longUrlRedirect(@PathVariable String shortUrl) throws URISyntaxException {
-        URI uri = new URI(urlService.startsWithHttpOrHttpsProtocolLongUrl(shortUrl));
-        HttpHeaders httpHeaders = new HttpHeaders();
-        httpHeaders.setLocation(uri);
-        return new ResponseEntity<>(httpHeaders, HttpStatus.MOVED_PERMANENTLY);
+        if(urlService.shortUrlExist(shortUrl)) {
+            URI uri = new URI(urlService.startsWithHttpOrHttpsProtocolLongUrl(shortUrl));
```

Comment:
Business logic shouldn't be in controller.
I think, we should check "url is exist" in service layer